### DISCUSSION TITLE
Existance of Hull() operation

### DISCUSSION THREAD

**Discussion by BarbourSmith** [sgenoud/replicad#77] [How To]
Does Replicad have a function similar to the OpenSCAD `hull()` operation?

https://spolearninglab.com/curriculum/lessonPlans/hacking/resources/software/3d/openscad/openscad_hull.html

I find that I use it quite a bit and I'm wondering if Replicad has something similar, or if there is some way that I could implement similar behavior.

**Comment by sgenoud** [comment]
As far as I can tell, `hull` operations are more of a mesh / CSG operation than a BRep one. Problems solved by hulls in openscad should be easier to solve with a `sweep` or a `loft` in replicad (or any BRep based tool).

The hull creates a convex shape around a set of points (and join them with lines). In a BRep environment we tend to have more complex structures than straight lines and it makes the whole endeavour more complicated.

This is as much as I understand. I have not used openscad myself that much, and therefore don't know directly how hulls are used. But I am open to see if I can do something if it is useful in practice!

**Reply by lf94** [reply]
Stitching together a curved surface is probably the closest thing. iirc open cascade can do this

**Comment by BarbourSmith** [comment]
That makes a lot of sense that it's more of a mesh based thing.

Here's a basic use case that I'm not sure how to accomplish with loft.

Let's say I have two extruded circles like this:

<img width="579" alt="image" src="https://user-images.githubusercontent.com/9359447/225329883-c7c160e5-365f-4da4-b7c6-40d5e9aca98b.png">

But I want them to be melded like this:

<img width="490" alt="image" src="https://user-images.githubusercontent.com/9359447/225330044-913de040-8ae0-4fea-aec0-8ad0b3ed947f.png">

How would I do that?

**Reply by lf94** [reply]
you would sketch using arcs and lines

**Reply by sgenoud** [reply]
This is [a quick implementation]( https://studio.replicad.xyz/workbench?code=UEsDBAoAAAAIADKJb1aVEp3EqwAAAE0BAAAHAAAAY29kZS5qc3WOsQrCMBCG93uKG1NJtRQEoTj4BiJuxSGmUQ%252FSRJKUDtJ3N7ahdLDb5bv%252F%252Fi%252FSGh%252FwA4DYONFzGPCITr01SdFUAHLc99SEV1yURQK%252BFVpfREOdj3if6J2eMzsAPDojA1mDrSDDssniVOicGWWszqfiHZZ8WXnLYhBxG15OqbMlE%252FzJSVZwzEvcLJORrL5SiSajrpbVC1U%252Bf3Td9BPNMf53TLdSW69YVsHwBVBLAQIUAAoAAAAIADKJb1aVEp3EqwAAAE0BAAAHAAAAAAAAAAAAAAAAAAAAAABjb2RlLmpzUEsFBgAAAAABAAEANQAAANAAAAAAAA%253D%253D) of what it could look like!

**Comment by avibryant** [comment]
This is a good short article on use cases of hull in OpenSCAD: https://hackaday.com/2018/02/13/openscad-tieing-it-together-with-hull/

**Reply by sgenoud** [reply]
Ok, this is definitely an interesting API to expose somehow - the workflow of having a bunch of shapes and asking to wrap them feels very useful.

**Comment by sgenoud** [comment]
It looks like [cad query has such a function](https://github.com/CadQuery/cadquery/blob/master/cadquery/hull.py) for 2D, but it only supports lines and arcs. I might give it a look (and see if it could be extended to cubic bézier as well).

**Reply by lf94** [reply]
as shown , the implementation with sketching is like 1000x simpler than the hull.py... i dont see why to bother.

**Reply by sgenoud** [reply]
Have you looked at the hackaday post? It feels to me that, while not as necessary as in the case of openscad a helper like `hull` can make some tasks easier to implement - it removes some geometry computations that the programmer would need to do.

**Reply by sgenoud** [reply]
Note that build123d has a [more straightforward (but approximative) approach](https://github.com/gumyr/build123d/blob/a38bc7ed55b93da161911777a99ec4370a4a410e/src/build123d/topology.py#L5659-L5771)

**Reply by lf94** [reply]
yes completely agree about the usage enabling less computations for a set of problems. i read and am familiar with hull. still clearly there are severe limitations to hull.py vs openscad's hull, which works on any arbitrary surface. we are entering brep limitations now. better to think of a diiferent workflow or operation to achieve similar developer performance.

**Comment by BarbourSmith** [comment]
For me the big selling point of hull over other options like using arcs and lines is that hull is easy to do with a flow interface, while arcs and lines work well with a graphical interface or a code interface.

**Reply by sgenoud** [reply]
It feels like there is also something to say about imperative vs declarative APIs. `hull` here feels like a declarative operation - you have a bunch of shape and you want to wrap them.

I guess that flow interfaces work better with declarative APIs (?). Similarly, do you miss a constraints API (a very declarative one)?

**Reply by BarbourSmith** [reply]
That's a great point about declarative vs imperative. The flow interface does seem like it works better to represent a declarative API. To some extent I think that's a bandwidth thing because the flow interface is somewhat limiting. I'm imagining that there will always be cases where the user will need to write Replicad code directly to accomplish certain things, but the more that they can do without needing to write code directly, the easier it is to pick up and get started.

**Comment by BarbourSmith** [comment]
What was the final verdict on this? I still feel like not having a hull option is really holding me back from making things the way that I'd like to

Does loft accept 3d shapes? That would do it for sure

**Reply by sgenoud** [reply]
I do not see hull as a 3D operation within replicad as opencascade does not support it and it would be a big investment to implement it.

 But I can see it as part of a 2D operations (where I can see myself implemententing such a functionality). Not sure if that would help you.

Loft works with 3D wires - I am not sure if that corresponds to what you mean by 3D shapes...

**Reply by BarbourSmith** [reply]
I think that a 2D version of Hull would meet 95% of what I'd be looking for. Especially if combined with loft to handle more of the 3D cases. 

A 3D wire is a path through points in 3D, right? It doesn't enclose a volume?

**Reply by BarbourSmith** [reply]
@sgenoud Any more thoughts on this? We're getting pretty close to having a flow type interface working, but this is the last big stumbling block that is preventing it from being truly usable. If you are still willing to implement a 2D operations version of this I would be happy to pay to get it done.

**Reply by sgenoud** [reply]
Sorry I have been quite busy, I will have more time from April on - is it ok if we talk about it then?

**Reply by BarbourSmith** [reply]
Absolutely, I totally understand :)

**Reply by BarbourSmith** [reply]
Did things end up slowing down 😛 ...They rarely do :)

**Reply by sgenoud** [reply]
April is not yet here - but it is close by!

**Reply by BarbourSmith** [reply]
Wise words 😂…I’m sorry I got excited 🫠

**Reply by BarbourSmith** [reply]
How is your schedule looking?

**Reply by sgenoud** [reply]
Better - I have started looking into it - it will take some time!

**Reply by BarbourSmith** [reply]
Thank you! Let me know if I can help (like pay money or test something or anything that helps!)

**Reply by sgenoud** [reply]
Not sure how I want to include this in replicad itself, but I have implemented a hull like operation (that I called `shrinkWrap`, because the hull is for points). 

<img width="843" alt="Capture d’écran 2024-05-31 à 15 18 20" src="https://github.com/sgenoud/replicad/assets/263325/e3ff6fa8-f7fa-419f-95ba-4ee8f1d532ef">


I have it [as loose code for now](https://studio.replicad.xyz/workbench#code=UEsDBAoAAAAIAKBpv1i1VTJoLwgAAMsZAAAHAAAAY29kZS5qc6UZaY8bNfR7foV3JaqZdnZyFCrIEgQUEEgFKrZSP0QRcjLOztC5sJ1NwpL%252FzrOf7bEns%252BGqVklsv%252Fu2O37%252BnHwpjy3L2JY8EvWr2ZKiahsuo2vO2rLY0Ow6JidiF2%252BKNXk%252BHo0QiDwnVJB8V5Zky5uKXOdStmI%252BHjNRpXxXj7%252BUeZHW9%252BN71lQ3Cu76djQaW7bk0SN7UnQ3TS0kiJJxuk%252FIN%252FBZ1PcJ%252BbrcsZYXtQRJFk4WRWpMcla2jIvRdldvZNHUpKIf2B27r1gtX%252B%252F4A4vaaULaWUweR4QgAwFUFAs4itOyqNm7JgKIWwDgTO54TUTasjoD5pqEWE5Wt6OTEY%252FT%252Bp4BhUgUf7CYLL4gX3FOj6kyQfRISlbfy3xO1Ck5JST6NSGFBiuAA5IoxB2t2NtGqQSElodJQo6TVUKWBxD2OIVfLWebQiiFFmTKbl5pCo%252BdhD9Smad0LaLDhNyQwzQmny88pGfPOoijgjj2IEAfMCCK80fRvi9k%252FhM7aHEo54ob8IJf8ElIKsDiLAIpb6Yx7lS0jTrVlgC5LEBs%252FU1ekOlqZQCBxYbKaKlAUvgBFBBssgIYdOK7vBDEeVCCBwWh2kXgAkLrDFb1rlozTiBCW2U3QcDtZKPcowHQLEJRo7WiT48e7B7UI6KpGKmYBMoUPqiUdJOzjMiGyJxVKeCiMD7UBqitWUBT7hvSrH9jGyBcbBUusgG3AqjcM1ZrGC2cMDam4i2KAgbObGBrtLeM6yhzHjYxUteg4MJaIdXru5y27NbBrG1iKLIaIPW2%252FvyTLPUmBK9D2RZcSC3KD3XGDoA36eiVtHcWCAhRNNWUSuax1kcerS5EO1HSbUnljypi3J7T1vJGawENB5LiliaJTFtrQjzpyOp1gttaFo98qDnltLIE0m6LScZvA3BtiwDa7SBwAN2q%252FV9MZehQbzyuPQxna0f%252FgYLmkUOOQ3lqyM7A1lGnrUq4mHxkzYIl6Al0Z71lSHHl5HOSpToNFmRpDoizqKarbe5taBvMO%252B2TPrDmMyeB3ANxFNB00TjvRWcHdUqelm%252FeKT4gqjvrRcJTkvd88K%252Bk7yXegPidB2xJrBte0RIaicAqY%252BUzJcu0oiXU5ekq8DbNaOuiN2qDbHD56SXEC9JCM%252B%252BC2MbOyYlkcLpQcN7oNDHFE6UC8VURFKBACRKv2YbuBCN7qKUlZzQ7kpxCNCo1tCC6iiuiaDVHNE1TTTDq16JZjE2o6GkXpCSo39kCYDFTxr3Ke3uGrHOxl5ltkJXqn963WeJLMJAdJuiSc6j%252FkBba1u8ZEZLCGEYlmYScvLjTKvu4J18D41atSLd%252Fir2YxA38Hqr6LxZDdccQxoqtthSJbuQQOVD58J7TNmiItWuS08mk1xBd8bet9AxNC6n6BCA8sMP3ajBd6Pk0vec0p9UddPMZBhICB4DdomssrZ1w%252FFZlHN5qz5sMAX9813ArpMzBJ5QzyGA5NBMoDBgdoiscHQuhvyNFL45dJNmca60XHCtw%252FZ5C64A6IGAalt4QQjHvjCDFGVtUYalzLiGCwTpbgTaKNyZUNdA825nqA%252B1SDantEqY7GyztzMZ%252FdRuK3c662LFyZ8zM1UTNeSBzwzMc6tQCHYAXCqgTNcPhTOwhSpyCvhZaiaAp6h0tUdrl062HggoHOLh1CanfezvTmIBZ2srykR9FmA%252BrS5RhHn8omp34W%252BowewGHM%252BIXOHYV1GftQs8cn5lwseir%252B%252ByZgT233WIxoIGGPo9ixE6Qo4toHyKMSVs0Rjbim7o8urD%252FwJgfFF3GDWabLgpq8TpXvURVEe%252FSE3UWNJeWbVFCm42iJV4eV0FGKOu1075X0Riz%252Fnbs9YUzxFRdjN%252Bpq7BCvn7zw0%252FfXsdBI7GtmpbCr8%252F9Vu7uelaArgNA5Fie3WZMroDh1LX5XpeX3HaDkzUIFMQIndGzo790doNNDhfwsHiiFzmrGujf2U7f4iXQiNB3bctoNyeYkgBewsubisl4IHF6EviJ6AtmEiOoBDBLeROxknfYp4bTxariXecDWp4fzml1h0HPvOokgyvclaPtJYRtolSFThjKntaXgvnKe4OwTyT%252B1X4w8u1cCn%252F7nHE9zbXF5gNWb7zAmcI9%252Br%252FJoqMFWfy2A2XVq4Dms6bCMBtMrJ7FJ%252BopZCArdJ0z08aNCpiBrAvk9rOvP%252FHiM4R%252BwyjLhKx30u9f4HPGQeZC%252FhOBPaEuiD4JBdb1TY0kQXXZlE3NonC6TY004bZRePOUku%252B9buzmDdRavaroYHDwJjgV3FdSjw0YSh4%252FG5PB1GruKe2ZwlfOj4NHnsX8OTf2%252BemYDumbab4vNj4dijNjGn0i8%252B1jooHMdYYz1ZOEKNYlmAhiVcx9wGlKfq6ZsaK68kCguF6GFyGQTPoos0so2CzPcF6m5N3eOEucIVVFloF0ESLHIbaKJWMFPRyhPfQT6N%252BE79WFfLs6D9oBJtPVYAQ%252Bwe%252BLAUaDLIaC3iNvH30R%252BKmm%252BNTDcjDlGtia7e3zdaR%252BuxfsSNfs2NyF%252FuFT%252BOuCb8BdvcdvdtBv8AiYsS3dlXjNFeYyKn7fwUB0V2RsTl5NVGpURf0Na9Xr9BTX9GDWs4%252FVGrISvA%252BrRItnOBjaxHtiL%252BrIf1DfoIALT9poCvdzCTd3AZMDiz5NyKtPh3qXfoc3jSd%252FU0C9Ujc%252FXD%252F01nh%252B84ldb6AvNNXrhtdQSl4GSDfTEGk2iPNJiPOyx%252Bizi0hQXTGiRr3XkUdULgFNy4bPybUoSgi%252Fa%252FvSY87nxmxPwnV3ZIyadAuDUYRIcQIu1Ld0%252FT8TfwFQSwECFAAKAAAACACgab9YtVUyaC8IAADLGQAABwAAAAAAAAAAAAAAAAAAAAAAY29kZS5qc1BLBQYAAAAAAQABADUAAABUCAAAAAA%253D), but you should be able to play with it (and find some bugs).

**Reply by sgenoud** [reply]
Ok, I packaged it here https://github.com/sgenoud/replicad-shrink-wrap, you can easily add it to a project in the studio (or import it via npm otherwise):

```js
import shrinkWrap from "https://cdn.jsdelivr.net/npm/replicad-shrink-wrap/dist/studio/replicad-shrink-wrap.js";
/** @typedef { typeof import("replicad") } replicadLib */
/** @type {replicadLib} */
const { drawRoundedRectangle, drawCircle, draw } = replicad;

export default function main() {
  const circle = drawCircle(7).translate(8, 18);
  const rect = drawRoundedRectangle(20, 14, 5);

  return [
    { shape: rect, color: "silver" },
    { shape: circle, color: "silver" },
    shrinkWrap(rect.fuse(circle), 50),
  ];
}
```

**Reply by BarbourSmith** [reply]
This looks AMAZING 😍!!!!

**Reply by BarbourSmith** [reply]
From some quick playing around it works great...also the updates to the workbench look phenomenal too!

Thank you!
